url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/utils/Optimization.java

```diff
@@ -0,0 +1,332 @@
+		//search(root, protocols, attributes);
+		DBProtocol.MAX_N = 2*DBProtocol.MAX_N;
+		protocols = DBProtocol.loadProtocols();
+		System.out.println("Total protocols: "+protocols.length);
+		searchByIteration(root, protocols, attributes);
+		
+	}
+
+	private static void search(Node root, DBProtocol[] protocols, Attribute[] attributes, State[] states){
+		List<DBProtocol> tmp;
+		Hashtable<String, String> names;
+		boolean[] stateStop = new boolean[states.length];
+		int removed = 0;
+		for (int i = 0; i < states.length; i++){
+			root.setState(states[i]);
+			BestProtocol best = new BestProtocol();
+			best.setNumberOfRounds(root.getTotalBitsExchangedDuringFastPhase()/2);
+			if (!best.dominate(root, attributes)){
+				//entonces nadie podra dominar a este hombre!
+				tmp = new LinkedList<>();
+			}
+			else {
+				System.out.println("The best succeed");
+				tmp = getNonDpominated(root, protocols, attributes);
+			}
+			names = getProtocolNames(tmp);
+			if (tmp.size() > 0){
+			//if (names.size() == 1){
+				System.out.println("One solution found, which is");
+				History.printInfo(root, attributes);
+				for (DBProtocol p : tmp) {
+					//History.printInfo(p, attributes);
+					System.out.print(p.getIdentifier()+", ");
+				}
+			}
+			if (tmp.size() > 0){				
+				stateStop[i] = true;
+				removed++;
+			}
+		}
+		System.out.println(removed*100d/states.length+" % removed: "+removed+ " out of "+states.length);
+		List<State> newStates = new LinkedList<>();
+		for (int i = 0; i < stateStop.length; i++){
+			if (!stateStop[i]){
+				root.setState(states[i]);
+				if (root.hasMafiaChildren()){
+					State s = states[i].clone();
+					s.moveToMafiaChild();
+					newStates.add(s);
+				}
+				if (root.hasDistanceChildren()){
+					State s = states[i].clone();
+					s.moveToDistanceChild();
+					newStates.add(s);
+				}
+				if (root.hasTerroristChildren()){
+					State s = states[i].clone();
+					s.moveToTerroristChild();
+					newStates.add(s);
+				}
+				/*if (root.hasBitsExchangedChildren()){
+					State s = states[i].clone();
+					s.moveToBitsExchangedChild();
+					newStates.add(s);
+				}*/
+				if (root.hasSizeOfMessageChildren()){
+					State s = states[i].clone();
+					s.moveToSizeOfMessagesChild();
+					newStates.add(s);
+				}
+				if (root.hasCryptoCallsChildren()){
+					State s = states[i].clone();
+					s.moveToCryptoCallsChild();
+					newStates.add(s);
+				}
+				if (root.hasMemoryChildren()){
+					State s = states[i].clone();
+					s.moveToMemoryChild();
+					newStates.add(s);
+				}
+				if (root.hasFinalSlowPhaseChildren()){
+					State s = states[i].clone();
+					s.moveToFinalSlowPhaseChild();
+					newStates.add(s);
+				}
+			}
+		}
+		State[] result = new State[newStates.size()];
+		newStates.toArray(result);
+		search(root, protocols, attributes, result);
+	}
+	
+	private static void search(Node root, DBProtocol[] protocols, Attribute[] attributes){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 157-174 into a separate method.